tests/types/ember-concurrency-test.ts

```diff
@@ -0,0 +1,1883 @@
+    }
+  });
+
+  test('rawTimeout', async () => {
+    expect<typeof rawTimeout>().toBeCallableWith(500);
+    expect<typeof rawTimeout>().parameters.toEqualTypeOf<[number]>();
+    expect<typeof rawTimeout>().returns.toEqualTypeOf<PromiseLike<void>>();
+
+    // @ts-expect-error
+    rawTimeout();
+
+    // @ts-expect-error
+    rawTimeout('nope');
+
+    // @ts-expect-error
+    rawTimeout(500, 'nope');
+
+    {
+      let result = await rawTimeout(500);
+      expect<typeof result>().toEqualTypeOf<void>();
+    }
+  });
+
+  test('waitForQueue', async () => {
+    expect<typeof waitForQueue>().toBeCallableWith('afterRender');
+    expect<typeof waitForQueue>().parameters.toEqualTypeOf<[string]>();
+    expect<typeof waitForQueue>().returns.toEqualTypeOf<PromiseLike<void>>();
+
+    // @ts-expect-error
+    waitForQueue();
+
+    // @ts-expect-error
+    waitForQueue(500);
+
+    // @ts-expect-error
+    waitForQueue('afterRender', 'nope');
+
+    {
+      let result = await waitForQueue('afterRender');
+      expect<typeof result>().toEqualTypeOf<void>();
+    }
+  });
+
+  test('waitForEvent', async () => {
+    type Evented = {
+      on(event: string, callback: (...args: any[]) => void): void;
+      off(event: string, callback: (...args: any[]) => void): void;
+    } | {
+      one(event: string, callback: (...args: any[]) => void): void;
+    } | {
+      addEventListener(event: string, callback: (...args: any[]) => void): void;
+      removeEventListener(event: string, callback: (...args: any[]) => void): void;
+    };
+
+    expect<typeof waitForEvent>().toBeCallableWith({
+      on(event: string, callback: Function): void {},
+      off(event: string, callback: Function): void {}
+    }, 'foo');
+    expect<typeof waitForEvent>().toBeCallableWith({
+      one(event: string, callback: Function): void {}
+    }, 'foo');
+    expect<typeof waitForEvent>().toBeCallableWith(
+      EmberObject.extend(Evented).create(),
+      'foo'
+    );
+    expect<typeof waitForEvent>().toBeCallableWith({
+      addEventListener(event: string, callback: Function): void {},
+      removeEventListener(event: string, callback: Function): void {}
+    }, 'foo');
+    expect<typeof waitForEvent>().toBeCallableWith(document.body, 'click');
+    expect<typeof waitForEvent>().parameters.toEqualTypeOf<[Evented, string]>();
+    expect<typeof waitForEvent>().returns.toEqualTypeOf<PromiseLike<void>>();
+
+    // @ts-expect-error
+    waitForEvent();
+
+    // @ts-expect-error
+    waitForEvent('nope');
+
+    // @ts-expect-error
+    waitForEvent(document.body, 'click', 'nope');
+
+    {
+      let result = await waitForEvent(document.body, 'click');
+      expect<typeof result>().toEqualTypeOf<void>();
+    }
+  });
+
+  test('waitForProperty', async () => {
+    let obj = { foo: 'foo' };
+
+    // @ts-expect-error (it actually works – bug in expect-type?)
+    expect<typeof waitForProperty>().toBeCallableWith(obj, 'foo', 'bar');
```

Comment:
1. TS (and thus `expectTypeOf`) is legitimately catching an error in the type as written, because of a quirk in how TS handles resolution of overloads. If you place the `value` overload *last* in this list, this works correctly. I poked at it a bit, and what you need for the best set of resolutions and inference is: 

    ```ts
    export function waitForProperty<O extends object, K extends keyof O>(
      object: O, key: K, callbackOrValue: (value: O[K]) => boolean
    ): Yieldable;
    export function waitForProperty(
      object: object, key: string, callbackOrValue: (value: unknown) => boolean
    ): Yieldable;
    export function waitForProperty<O extends object, K extends keyof O>(
      object: O, key: K, callbackOrValue: O[K]
    ): Yieldable;
    ```
    
    If you do that, you'll be able to drop the `@ts-expect-error` on line 1561, and you'll also need to update line 1565 because the parameters resolved for the default overload will be different, but you'll no longer have these inference errors.

2. Marking this as an error to be ignored means that your *consumers* will have to do the same! Any time you hit a thing like this, feel free to ping one of the Typed Ember folks for input—ending up with something that consumers of the types have to work around as a cast is always 😬.

Reply:
Thanks for digging into this. I don't (or at least didn't) think the previous version was _actually_ a problem – that's why this line was followed by an actual call to the function to confirm that it works w/o type errors. Either way, it's good to have it fixed and thanks for the pointers.

Reply:
Ah, I see what you meant by the original comment now. You're correct that it was _callable_ the other way, and didn't show a type error, but when I inspected it directly it was resolving the wrong overload. 😬 Kind of weird, to be honest, but in this case I think `expectTypeOf` was catching something that likely would have shown up in a *really* weird way in the wild.